lib/redshift/client/connection.rb

```diff
@@ -11,16 +11,17 @@ def initialize(configuration)
         @original = PG.connect(configuration.params)
       end
 
-      def_delegators \
-        :@original,
-        :exec,
-        :exec_params,
-        :escape,
-        :escape_string,
-        :escape_literal,
-        :close,
-        :transaction,
-        :quote_ident
+      def method_missing(method_name, *args, &block)
```

Comment:
So `def_delegators` uses `method_missing` in the background? I don't see, how `method_missing` will be invoked, but the method itself looks good.

Reply:
`def_delegator` is a method of Forwardable module. It is simply delegating method calls to a receiver. It is part of ruby library. The bottleneck here is that you have a set of methods what you are able to delegate and need to maintain the list.

In ruby if you call a `:method`, then it try to call it on the instance's class, if it is not found then it tries to lookup on the ancestors of the class. On default `method_missing` is a ruby method which will be called whenever a method is not found in the ancestor tree and it will raise maybe a NoMethodError. But you can override this behaviour and delegate this decision to any object which in our case the PG::Connection. So from now on this object will decide whether it is able to handle the commands or not. So it is kinda like a rescue method for method calls.